### ISSUE TITLE
idXML, featureXML, consensusXML schema extension

### ISSUE THREAD

**Issue by timosachsenberg** [OpenMS/OpenMS#1671]
It has been quite a while since the last schema update and I think we should take the effort to improve on the current state. I am willing to do the update if we can agree on on the extension.
Questionable (though practical) decisions like storing small molecule identifications in peptide/protein structures could be corrected, missing information like the position of a peptide in a protein for shared peptides or the length of a protein (e.g. used for coverage calculations) are potential issues we can tackle.

Please feel free to add your suggestions there so we can discuss the implications.
By the way, changes should be backward read-compatible (= we are able to read old files in newer version.)
This issue is intended to collect all relevant issues that might be resolved by a schema update as well as to channel the discussion.
#### #1031 Protein coverage not properly calculated

depends on: #999 #140
#### #999 PeptideIndexer should provide option to annotate AABefore/AAAfter and location in protein in PeptideHits

partially resolved in merged PRs #1048 #1078
resolved in: #1666 #1667 

by storing aa_before, aa_after, start, end for every peptide evidence
#### #140 ProteinHit::coverage_ not stored in file formats

TODO
#### Store small molecule identifications in idXML, featureXML, consensusXML without hacking them into peptide/protein sturctures

proposal in: #1659
#### #1435 ProteinIdentification::SearchParameters need update for relative tolerances

partially resolved in merged PRs: #1471
#### Store group type in ProteinGroups

TODO
#### Extend enzyme element to accept arbitrary enzyme names

resolved in https://github.com/OpenMS/OpenMS/pull/1675
#### use same capitalization of UserParam in all file formats (replace "userParam" in featureXML/consensusXML)

resolved in https://github.com/OpenMS/OpenMS/pull/1675

**Comment by hendrikweisser** [comment]
I can't find an open ticket for it, but this has caused problems several times: Peptide IDs in consensus maps are always attached to the consensus feature, never to a sub-feature. If multiple feature maps annotated with peptide IDs are linked together into a consensus map, the provenience information - from which map each ID came - is lost.
(Attaching IDs directly to consensus features can still be useful, e.g. if IDMapper is run directly on a consensus map.)

**Comment by timosachsenberg** [comment]
Sounds like a good extension to allow IDs at sub features.
See the last few lines of:
```
        <xs:element name="element">
                <xs:complexType>
                        <xs:attribute name="map" type="xs:unsignedInt" use="required">
                                <xs:annotation>
                                        <xs:documentation>The element's original map index.</xs:documentation>
                                </xs:annotation>
                        </xs:attribute>
                        <xs:attribute name="id" type="xs:string" use="required">
                                <xs:annotation>
                                        <xs:documentation>The element's unique id in its original map.</xs:documentation>
                                </xs:annotation>
                        </xs:attribute>
                        <xs:attribute name="rt" type="xs:double" use="required">
                                <xs:annotation>
                                        <xs:documentation>Retention time of the element.</xs:documentation>
                                </xs:annotation>
                        </xs:attribute>
                        <xs:attribute name="mz" type="xs:double" use="required">
                                <xs:annotation>
                                        <xs:documentation>M/Z of the element.</xs:documentation>
                                </xs:annotation>
                        </xs:attribute>
                        <xs:attribute name="it" type="xs:double" use="required">
                                <xs:annotation>
                                        <xs:documentation>The intensity of the element.</xs:documentation>
                                </xs:annotation>
                        </xs:attribute>
                        <xs:attribute name="charge" type="xs:integer">
                                <xs:annotation>
                                        <xs:documentation>The charge of the element. A missing change attribute is equal to charge 0.</xs:documentation>
                                </xs:annotation>
                        </xs:attribute>
                        <xs:sequence>
                            <xs:element name="PeptideIdentification" type="PeptideIdentificationType" minOccurs="0" maxOccurs="unbounded"/>
                         </xs:sequence>
                </xs:complexType>
        </xs:element>
```

**Comment by hendrikweisser** [comment]
> Sounds like a good extension to allow IDs at sub features.

Yes, although this would also require changes in the library, as currently a FeatureHandle (representation of a sub-feature) can't store peptide IDs. (It's all a bit of a mess.)

As an alternative, would it be possible to store the corresponding map index together with a peptide ID? And would this be better or worse?

**Comment by timosachsenberg** [comment]
Upgrading a FeatureHandle to store PeptideIds would be the cleaner solution in my view.

**Comment by hroest** [comment]
just a question: how would metabolite ids be stored in consensusXML format, would there be a separate tag?

@timosachsenberg when updating the FeatureHandle please consider i) that we also would like to store metabolites and ii) that it should not use up a lot of memory when there are no metabolite or peptide ids

**Comment by timosachsenberg** [comment]
Yes, both PeptideIdentification and SmallMoleculeIdentification.
I have to see how we do this memory efficiently but we could always add a pointer to a struct that holds a vector of peptideidentification/smallmoleculeidentification without wasting much space.

**Comment by timosachsenberg** [comment]
we had some discussion in our MS group today and some asked whether we really need this right now. The argument was, that we could use the featureXML files together with the consensusXML file right now if we really need to this information.

**Comment by hendrikweisser** [comment]
> we had some discussion in our MS group today and some asked whether we really need this right now. The argument was, that we could use the featureXML files together with the consensusXML file right now if we really need to this information.

I don't think this is a very good argument. You can find work-arounds for almost every issue. In this concrete case, adding meta values (for the map indexes) to the peptide IDs would be my preferred work-around.

In my opinion, the better question would be if we shouldn't just make the switch to mzIdentML and mzQuantML instead.

**Comment by timosachsenberg** [comment]
having had a deeper look at mzIdentML I don't think it plays well as intermediate format. I think it is still useful for reporting the final results but otherwise to clumsy.

**Comment by hendrikweisser** [comment]
> having had a deeper look at mzIdentML I don't think it plays well as intermediate format. I think it is still useful for reporting the final results but otherwise to clumsy.

Woah, wait a minute. I was under the impression that idXML was supposed to be phased out in favour of mzIdentML as soon as possible. Has there been a discussion about this that I wasn't part of (e.g. at the user meeting)?

I thought the whole point of mzIdentML was to have a format that doesn't suffer from the limitations of its predecessors (idXML, pepXML, ...). In other words, if you tried to fix all the problems of idXML, you'd basically end up with mzIdentML.

**Comment by timosachsenberg** [comment]
I think 100% support for mzIdentML is one of our primary goals. But even with 100% support it doesn't solve every problem (e.g. small molecules) and we will not get rid of idXML on a medium time frame.
The decision if we can phase out idXML will therefor be postponed after we gathered enough experience with mzIdentML. Anything else is imho not practical given the resources we have right now (i.e. we simple don't have enough time right now to extend the mzIdentML standard to small molecules and do the OpenMS changes now).

**Comment by hendrikweisser** [comment]
I have two problems with this:
1. We are investing time extending idXML to cover things that are already solved in mzIdentML.
2. Adding new capabilities (e.g. small molecule support) to idXML means double the work, because mzIdentML will have to be extended to cover that anyway. (Unless we want a whole new file format, which we probably don't.)

I think mzIdentML will not see heavy use within OpenMS until we phase out idXML. I'd say as soon as #1548 is solved, we should move to replace idXML with mzIdentML - in order to not have to make lots of fixes to idXML that will only become obsolete shortly after.

As for the small molecules, my suggestion would be to propose an unofficial extension to mzIdentML and implement that in OpenMS - with a view of having it become part of the standard, but not caring too much about obtaining consensus at this point (i.e. just make something that works for us - same as we would do with idXML).

I don't understand why extending mzIdentML to small molecules would take more time than extending idXML. (That's not even considering that the extension of mzIdentML would still have to happen later!)

**Comment by lars20070** [comment]
> On 26 Oct 2015, at 11:14, hendrikweisser <notifications@github.com> wrote:
> 
> I can't find an open ticket for it, but this has caused problems several times: Peptide IDs in consensus maps are always attached to the consensus feature, never to a sub-feature. If multiple feature maps annotated with peptide IDs are linked together into a consensus map, the provenience information - from which map each ID came - is lost.

+1 Better to attach IDs directly to sub-features.

The current work-around is to switch on `annotate_ids_with_sublements` in `IDMapper`. The map index is then a metavalue of the id.

**Comment by timosachsenberg** [comment]
My general feeling is, that the extensive cross-references in mzIdentML will make simple things like e.g. manually creating a test file difficult. Adapting idXML (let's call it partially converging to mzIdentML) will most likely result  in a simpler file format that covers everything we need while beeing less complex (as it doesn't need to cover everything we don't need). That we want to export to mzIdentML is clear - it might be just easier, faster, less memory consuming (recall that we still use dom for mzIdentML) and less complex to do this with our own format. And we are just talking about idXML right now - which is basically not much work to update anyways. For mzQuantML things look much more difficult.

**Comment by hendrikweisser** [comment]
> My general feeling is, that the extensive cross-references in mzIdentML will make simple things like e.g. manually creating a test file difficult.

Okay, that's a valid point. But consider this:
- If you extend mzIdentML to small molecules, you could publish that. (I doubt you could publish changes to idXML.)
- Having small molecule support in idXML, but not in mzIdentML, would effectively tie us into keeping idXML as the main ID format in OpenMS.

> That we want to export to mzIdentML is clear - it might be just easier, faster, less memory consuming (recall that we still use dom for mzIdentML) and less complex to do this with our own format.

The problem is that we can't capture all the information that we want in a complete mzIdentML in our own format - unless we basically turn it into mzIdentML.

**Comment by hroest** [comment]
I agree with @timosachsenberg that until we have a fast and memory-efficient implementation of mzIdentML it does not make sense to switch, otherwise we just run into scalability problems. However I agree with Hendrik that we should, if possible, make sure that our internal datastructures are agnostic to the file format so that switching would be easier in the future.

@timosachsenberg I understood your comment "partially converging to mzIdentML" in that respect, e.g. that the internal data structures will be easier to map to each other. Is this what you meant? I am all for that...

For small molecules, I feel the main issue is whether it is faster/easier to extend mzIdentML or idXML. Also, I am not sure whether producing non-standard conform mzIdentML is really the best way to go here.
